lib/streamlit/type_util.py

```diff
@@ -712,6 +714,60 @@ def is_pyarrow_version_less_than(v: str) -> bool:
     return version.parse(pa.__version__) < version.parse(v)
 
 
+def _maybe_truncate_table(
+    table: pa.Table, truncated_rows: int | None = None
+) -> pa.Table:
+    """Experimental feature to automatically truncate tables that
+    are larger than the maximum allowed message size. It needs to be enabled
+    via the server.enableDataframeTruncation config option.
+
+    Parameters
+    ----------
+    table : pyarrow.Table
+        A table to truncate.
+
+    truncated_rows : int or None
+        The number of rows that have been truncated so far. This is used by
+        the recursion logic to keep track of the total number of truncated
+        rows.
+
+    """
+
+    if config.get_option("server.enableDataframeTruncation") is not True:
+        return table
+
+    # The maximum size allowed for protobuf messages in bytes:
+    max_message_size = int(config.get_option("server.maxMessageSize")) * int(1e6)
+    # We add 1 MB for other overhead related to the protobuf message.
+    # This is a very conservative estimate, but it should be good enough.
+    table_size = int(table.nbytes) + 1 * int(1e6)
+    table_rows = table.num_rows
+
+    if table_rows > 1 and table_size > max_message_size:
+        # Calculate an approximation of how many rows we need to truncate to.
+        targeted_rows = math.ceil(table_rows * (max_message_size / table_size))
+        # The targeted_rows is just an approximation. We cut out another
+        # 5% of the number of truncated rows to make sure that we don't
+        # execute this truncation logic too many times since it is quite
+        # inefficient and its running in a recursion.
+        rows_overhead = math.floor((table_rows - targeted_rows) * 0.05)
+        # Make sure to cut out at least a couple of rows to avoid infinite recursion.
+        targeted_rows = max(min(targeted_rows - rows_overhead, table_rows - 5), 1)
+        sliced_table = table.slice(0, targeted_rows)
+        return _maybe_truncate_table(
+            sliced_table, (truncated_rows or 0) + (table_rows - targeted_rows)
+        )
+
+    if truncated_rows:
+        st.caption(
+            f"⚠️ Showing only {string_util.simplify_number(table.num_rows)} rows "
+            f"from a total of {string_util.simplify_number(table.num_rows + truncated_rows)} "
+            "to comply with data size limitations."
+        )
```

Comment:
Hi @LukasMasuch 👋 

@sfc-gh-xhe asked me to look at this PR from a design perspective, because it's possible that we may use this in Notebooks too. 

Did Jessi Shamis happen to review this design already? If not, no worries.

A few thoughts on this message:

- It might be good to shorten this message by cutting out a few words. For example, we could say "Showing X out of Y rows due to data size limitations."
- I don't quite understand why we're saying "comply with". Is there some kind of legal/security constraint that is limiting us from showing all the rows?

Reply:
I updated the suggested text 👍  in a way, the underlying limitation is a security constraint to prevent apps from taking down GS by sending through too much traffic, but the `comply with` indeed sounds a bit harsh.